### ISSUE TITLE
Planning issue for translation/internationalization/localization systems

### ISSUE THREAD

**Issue by jywarren** [publiclab/plots2#3262]
We have some basic internationalization/localization via the `i18n` gem, but need to think through how to improve it, and ALSO come up with a streamlined workflow to recruit and integrate new translations on an ongoing basis. This is a planning issue for this set of projects.

****

## Translation systems

Our translation files are spread over lots of files and folders in https://github.com/publiclab/plots2/tree/master/config/locales

I'm inclined to make them all one big file, or at least consolidate them some. One folder with key/value would be SO AWESOME! I wasn't sure why our implementation fragmented everything across so many folders. I think we could do this in Transifex.

Discussion of alternative platforms via @ebarry :

* LocalApp: https://www.localeapp.com/projects/9741/translations/5689967 
* https://crowdin.com/ (github integration)
* https://poeditor.com/ (github integration)

## Ongoing translation workflow

We need a workflow to regularly go from Transifex (or something similar) back into the translation files. 

Transifex has GitHub integration: https://docs.transifex.com/integrations/github/ and awesome ability to auto-sync in TravisCI: https://docs.transifex.com/integrations/github/#integrating-the-client-with-travis-ci

This would grab new translations each time we accepted a pull request, and register new texts that need translation!

### Outreach

For actually doing the translations, I'm hesitant to make a huge # of translation issues in the same repository, even if they are first-timer type issues. I think we could maybe get more contributors to translate outside the GitHub interface (which is most welcoming to coders). 

I think we'll need a good outreach game to actually recruit people to do translations! Maybe, we could tweet them out? We could show a random phrase at the bottom of the site each time the site is loaded, that needs translation?

****

If we happen to find alternatives (especially open source ones) that's great, but we can stick with Transifex if there's no big reason to switch. 

The ideal might be that with fewer files, we could copy in new changes from either another service like Transifex periodically (monthly?), or from files in another repository (although this doesn't make it more accessible by going outside the GitHub interface)

We might even try to automate the import and run it periodically. But it could be a regular task manually too. 

## Translation prompts

Anywhere we display translated text using the `t('some_string_here')` helper, we ought to display a little globe icon with a tooltip saying `Needs translation? Click to help translate this text.` or something (ideally in the browser-configured language!)

## Translating content

We'll need to think about how to translate both regular site content (wiki pages + notes) and ALSO features, which are used around the site for things like the static footer, explanatory text, etc.

Perhaps there should be a language type on revisions, and we could make it possible to track translated revisions alongside regular ones. Or perhaps a translated node, linked to the main node?

Compare https://publiclab.org/barnraising to https://publiclab.org/wiki/barnraising-guangzhou-china for an example of 2 versions of a page in 2 languages. 

### Features

Some parts of site content -- like the footer and front page -- are run in nodes, as database content. Let's think of a way to have features like `frontpage-en` vs `frontpage-zh` and `frontpage-es` 

## Helpers for translation

@siaw23 points out, and I agree, that translation is tough for newcomers submitting new code. The current format is something like:

```ruby
<%= t('some_string_here') %>
```

...where `some_string_here` is the key to some string of text. 

I propose a new helper, maybe called `translate()`, which looks for translations, but if none are present, displays the key text itself. That way, we can create new interfaces that include text, without first having to do translation for them -- they'll have the fallback text. So:

```ruby
<%= translate('Click here to begin') %>
```

which initially would have only the default english, but via the prompts (above) would encourage community translation efforts. 

## Scanning for translation opportunities

Finally, we might have a script which scans for uses of the new helper (from above) that don't have entries in the default translation file, and copies them in. 

****

This is a lot. Just kicking off some discussion and we can start to move things into separate issues soon! Love to hear peoples' input!

**Comment by jywarren** [comment]
I've broken out the "move translations" issue into a multiple-use issue good for newcomers! https://github.com/publiclab/plots2/issues/3515

**Comment by gauravano** [comment]
Hi @jywarren @ebarry , I went through the details of the project and done some research at my end. 

I saw that Transifex is being considered right now for integration, but what I that was paid. There are two more methods which I am considering:
**1. Using Copycopter gem**
In this, the web interface would be hosted as a separate application and can be hosted via Heroku too as it only requires one dyno to run.

Although, the project is Abandoned some time back but, it can do our work. For details, check here - https://github.com/copycopter/copycopter-server

**2. i18n backend**
It is one of the solutions which offers control of the system in our hand. Please check details here- http://railscasts.com/episodes/256-i18n-backends?autoplay=true

I prefer the second option as with that we wouldn't have to depend on other services and also can use the interface of our preference.

Thanks!

**Comment by jywarren** [comment]
In the second option, what backend would we use? Not flat yml files? How
would people edit them then?

Is it possible to use Transifex or another management system to organize
the crowdsourcing of translations, so we could kind of quickly get people
plugged int translating short strings?

Thanks Gaurav!

**Comment by gauravano** [comment]
@jywarren @ebarry I am recording the Translation Project progress at https://github.com/publiclab/plots2/projects/9 . Thanks!

**Comment by gauravano** [comment]
@jywarren @ebarry can I create a test project on transifex with publiclab's URL for testing? Thanks!

**Comment by jywarren** [comment]
Hi, Gaurav - thanks - i watched the video and perhaps we should open a new issue for this discussion if it gets much longer? 

In any case, the feature of Transifex is not the auto-translation, but the idea that people might show up and want to help, but we might want to have translations reviewed, moderated, and managed. The language in this post is a bit overly commercial for us, but some of the principles are the same:

https://www.transifex.com/blog/2016/crowdsourcing-translations-and-moderation-programs/

https://docs.transifex.com/localization-tips-workflows/crowdsourcing-translations has great guidelines for this. The experience for people who offer to translate looks pretty smooth and welcoming, which is nice esp. if many translators aren't coders. And there are different roles: Language Coordinator, Reviewer, and Translator. 

What do you think of this? I believe we can do integration as noted above such that the translation texts are under our control and in standard formats. 

But the pricing is rough. $140 per month... 😦  @ebarry did you know this? https://www.transifex.com/pricing/

I wonder if there are other options that have some of the desired functions:

1. easy interface and login for volunteer translators
2. ability to review translations
3. maybe ability to have translations "approved" by a second person (reviewer?)
4. way to use a POST request to link into a specific translation if it doesn't exist, with the text to be translated, so that it will drop the resulting translation into our locale files???

**Comment by jywarren** [comment]
And my concern is that Redis and Copycopter don't have the interface or review parts... hmm.

**Comment by jywarren** [comment]
Hmm, https://github.com/topics/i18n ?

https://weblate.org/en/hosting/ is based on open source but much cheaper than Transifex

https://github.com/parrot-translate/parrot seems unsupported

**Comment by jywarren** [comment]
Ooh this is nice though! https://github.com/glebm/i18n-tasks -- can find unused translation strings and help manage them in a CLI!

**Comment by jywarren** [comment]
http://pootle.translatehouse.org/

https://www.accent.reviews/

Hmm interface here is pretty complex: https://demo.weblate.org/translate/hello/master/fr/?type=unapproved&offset=2

**Comment by jywarren** [comment]
Whoa.... https://github.com/microsoft/localization/wiki/How-to-contribute-to-Software

**Comment by jywarren** [comment]
https://crowdin.com/pricing#annual is $16/month

**Comment by jywarren** [comment]
And POEditor has a free tier that might work for us: https://poeditor.com/pricing/

**Comment by jywarren** [comment]
I just used localeapp for a while and it's pretty nice. It supports only 2 locales on free tier though. https://www.localeapp.com/pricing

@ebarry can we afford $24/month for 5 locales at some point soon? Or should we try something else. At least this looks easy...

**Comment by jywarren** [comment]
OK - sorry for so many comments. Let's give POEditor a try based on its pricing and nice interface. What do you think, @gauravano ?

![screen shot 2019-01-09 at 6 21 30 pm](https://user-images.githubusercontent.com/24359/50935859-7b8d6200-143b-11e9-9e0b-c29d69bbae14.png)

**Comment by gauravano** [comment]
> Hi, Gaurav - thanks - i watched the video and perhaps we should open a new issue for this discussion if it gets much longer?

@jywarren the backend option is one where we have to do the most work, like prepare the translation page, and you mentioned we also want reviewer features too, so it would take time. Using this option, we can open the modal instead at same page and take user input, but at same time we have to put the moderation system in place too. We can integrate this, but would consume time.So, if you want I am open for discussion on this. Let me know, what do you think.


> In any case, the feature of Transifex is not the auto-translation, but the idea that people might show up and want to help, but we might want to have translations reviewed, moderated, and managed. The language in this post is a bit overly commercial for us, but some of the principles are the same:
https://www.transifex.com/blog/2016/crowdsourcing-translations-and-moderation-programs/
https://docs.transifex.com/localization-tips-workflows/crowdsourcing-translations has great guidelines for this. The experience for people who offer to translate looks pretty smooth and welcoming, which is nice esp. if many translators aren't coders. And there are different roles: Language Coordinator, Reviewer, and Translator.

Yes, transifex provides many features and I have signed-up on their platform and trying various features and also sent a invite to you Jeff. I am sending one to you to @ebarry.  

@jywarren @ebarry these are free services too, offered by platform, like we can work on file-based project/system, where we have to push the changes to yml files. But, reviewing, inviting translators, etc is free. Yes, they have to sign-in to transifex to translate the string and their request for acting like translator can be reviewed or automatically accepted. Try out and see if free plan is sufficient for us. 

> I wonder if there are other options that have some of the desired functions:
1.easy interface and login for volunteer translators
2.ability to review translations
3.maybe ability to have translations "approved" by a second person (reviewer?)
4.way to use a POST request to link into a specific translation if it doesn't exist, with the text to be translated, so that it will drop the resulting translation into our locale files???

Transifex offers 1-3 for free but I can't say about 4. We can embed URL like this for missing word - https://www.transifex.com/own-20/publiclab-test/translate/#de/enyml/161766077?q=text%3AEnglish
 and the user/translator would be redirected to the word after login. But, I am unable to find any sync method which is free till now, so we have to manage the timely push/pull ourself if we decide to ride on free plan.

**Comment by gauravano** [comment]
> And my concern is that Redis and Copycopter don't have the interface or review parts... hmm.

Yes, in both of these options we have to make the review system if we want to use them.

**Comment by gauravano** [comment]
Finally, gone through all your suggestions :smile:. I have signed-up for POEditor and would invite you too to explore. Also, it provide free support to Open Source projects:

![poe_editor](https://user-images.githubusercontent.com/20878070/50966180-2fedbd80-14fa-11e9-8cde-1a92b9acadd7.png)

Thanks!

**Comment by jywarren** [comment]
Ah, we are already set up as an open source org! Great.

**Comment by jywarren** [comment]
OMG this was a long time ago, embarrassing. Glad we're getting back into it! Travis integration sounds great. I get how it will update Transifex with new strings. But how would it fetch Transifex output... would i have to make a commit from Travis?

**Comment by gauravano** [comment]
Wow, the PublicLab's transifex account seems active from a long time.

**Comment by gauravano** [comment]
@jywarren at transifex account, the whole dataset is present in categories and there is lot of translations available, so what would be our step regarding making a single file for all translation?

I also saw this - 
`Commit the travis.yml file to the branch you want to sync translations from. Now every time you commit changes in this branch your content will be uploaded to Transifex and your resources will be updated automatically.` 

I think we can try integrating with travis. What do you say?

**Comment by gauravano** [comment]
I have started working on the helper function part and have some queries:

>@siaw23 points out, and I agree, that translation is tough for newcomers submitting new code. The current format is something like:
`<%= t('some_string_here') %>`
...where some_string_here is the key to some string of text.
I propose a new helper, maybe called translate(), which looks for translations, but if none are present, displays the key text itself. That way, we can create new interfaces that include text, without first having to do translation for them -- they'll have the fallback text. So:
`<%= translate('Click here to begin') %>`
which initially would have only the default english, but via the prompts (above) would encourage community translation efforts.

So, we want to make a helper function which will take English string as argument instead of I18n key, and also take various options required in `<%= t() %>` as our helper function would need those options too.

Am I getting it right @jywarren?

**Comment by gauravano** [comment]
As all the locale files have one thing in common i.e., key so we need to find a way to find the key from the string value. And, after that we could use translate helper function as it is passing in the key with options.

**Comment by jywarren** [comment]
maybe we could use the English string, in quotation marks, as the key?

Great!

**Comment by jywarren** [comment]
Or alternatively we could periodically run  https://github.com/glebm/i18n-tasks

**Comment by gauravano** [comment]
I have started with Translation prompts and here's the progress:
![translate3](https://user-images.githubusercontent.com/20878070/51215514-2e335800-1947-11e9-85b3-e2fdae96930d.gif)

Suggestions?

**Comment by jywarren** [comment]
Cool! Are the icons in `<i class="fa fa-globe"></i>` format? I think that'd make them a bit more inline, vertically aligned. We could make them quite light, like `#bbb` color maybe? And a Bootstrap tooltip style is a little nicer looking than the default tooltip.

This is looking really cool though! Sometimes I wonder if very tiny words like `by` may not be worth it, as it kind of disrupts the flow; but maybe those would be easiest to auto-translate or us existing lookups, and so they wouldn't appear for most languages?

This is looking great, Gaurav!

**Comment by gauravano** [comment]
> Cool! Are the icons in <i class="fa fa-globe"></i>
Yes, it's font-awesome's globe icon. Actually, I intentionally added some spacing so I can remove that. I have added a helper method which replaces translation with a span tag and that tag contains link, icon and tooltip. So, this way we would be able to control all the translations at once. And, we can add a condition like if translated string is less than ___ then, don't show icon. 

Any comment on this - https://github.com/publiclab/plots2/issues/3262#issuecomment-454571219 @jywarren ?

Thanks!

**Comment by jywarren** [comment]
> they resemble Public lab's file structure.

Yes, i think they were just from a much older version of the site.

> Are we sure about deprecating the file structure of locales and adopting single file structure?

The reasoning had been:

1. easier for newcomers to find a string, and shortens the lookup string when using `t(...)` - i.e. we can use english strings as the lookup keys in a flat file without needing subfolders to be in the string
2. less redundancy if we use the same phrase multiple places
3. easier to skim through to check a lot of translations in a single language, potentially for coders who speak that language to make a lot of quick changes, or for other reasons
4. easier to generate issues for, since we could prompt people to input into a fixed file location, rather than have to guide people on making folders/subfolders, naming convention, etc etc.

But we should examine the cons/drawbacks as well. Does it create trouble for... merging? Transifex integration?

> We want to automate the process, I want to try import and export using Public Lab's repo. Can I make a new project and try it there, so that we don't touch present files at Transifex?

Yes, for sure. You could also try maybe importing the Transifex strings into a repo, compressing them into one file, then updating Transifex from it? I dunno, i mean, how Transifex tracks moving .yml files is unknown to me. 

> After making language specific files, the import would divide all the string in categories? How we are going to handle the export?

Import in which direction? In Transifex categories? I wonder... maybe as we grow, we'd then later want to begin making subfolders, but only when needed? 

Here's an idea - what if we, in the translation helper(s), could detect where in the site the string is being used, from the controller, action, or maybe just the URL (so it could be consistently implemented in ruby and JS), and auto-append subfolder prefixes based on that? It's a bit more complex and lacks some of the simplicity of 1) above, but maybe we could write out the idea for use later once the translation files get longer and really need categories?

A lot hangs on the question of how Transifex deals with locale files which move in GitHub, and if it knows how to deal with or track (or even ignore) strings getting moved around.

**Comment by jywarren** [comment]
Yep, sorry took me a bit to think through and write it out! Thanks @gauravano !!!

**Comment by gauravano** [comment]
No issue :sweat_smile:

**Comment by gauravano** [comment]
> 1. easier for newcomers to find a string, and shortens the lookup string when using t(...) - i.e. we can use english strings as the lookup keys in a flat file without needing subfolders to be in the string

Suppose we combine all the folders and generate 3 yml files of specific languages. Then, those files would be containing keys in nested hash-value form like - https://github.com/svenfuchs/rails-i18n/blob/master/rails/locale/en.yml (this is the form we currently use). If the yml file would exist in this form then we would be retrieving translation using `t(dashboard.dashboard.by)` and so on because keys are in nested form. 

If we want to show user-friendly text to the contributors like `translation("This is a dashboard alert")` then, 
there are 2 ways:
1. Storing the key-value directly, without nesting - This would require changing all the yml files keys and also assigning user-friendly keys to all the methods.
2. Currently, we use SimpleBackend for storing translations which comes by default with rails, in this all yml files and their translations are loaded in in-memory hash - https://www.rubydoc.info/github/svenfuchs/i18n/I18n/Backend/Simple/Implementation

Suppose we want to show user `translation("Hello, this is welcome text")` --> This is friendly I believe, although for me using nested key is also friendly

So, all the keys which are present in the files would remain as same in nested form. The text "Hello, this is welcome text" will be passed onto the helper. This text would be the English translation of the the nested key and I would find the key using recursively iterating over I18n.backend.translations[:en] --- something like that. So, we would be finding the key for the the t occurances but that is necessary as key is a common link in-between all languages. The problem I am predicting here would be the use of variables as translation also use variables for making the value. 

> 3. easier to skim through to check a lot of translations in a single language, potentially for coders who speak that language to make a lot of quick changes, or for other reasons

Right!

> 4. easier to generate issues for, since we could prompt people to input into a fixed file location, rather than have to guide people on making folders/subfolders, naming convention, etc etc.
But we should examine the cons/drawbacks as well. Does it create trouble for... merging? Transifex integration?

Taking current scenario in perspective, we have to combine all the subfolders in a single file if we decide on this one. Transifex divide the keys in categories so all making single file we would be still having the keys in nested form so if we download the translated file from Transifex, it would export in a separate file. I will check if Travis integration can make it any easy for us.

> A lot hangs on the question of how Transifex deals with locale files which move in GitHub, and if it knows how to deal with or track (or even ignore) strings getting moved around.

yes! That's the thing. I try to integrate Travis on my fork and replicate the integration. 

@jywarren

**Comment by gauravano** [comment]
I gone through this doc - https://docs.transifex.com/projects/updating-content/#section-automatic-updates which lists various ways to import the translation files. Currently, the translation files present at transifex got there by uploading from a local device(someone's computer). But, for long run we would be using Auto-update method in which we would add URL to our files.

**Comment by jywarren** [comment]
OK first, wow, this is a great idea 🎉 , and the helper really lets us do things like this:

> This text would be the English translation of the the nested key and I would find the key using recursively iterating over I18n.backend.translations[:en] --- something like that. 

Re variables, yeah, i see that. Could we include variables in identical format and parse them out somehow? I dunno... i mean, the variables could still be in the english strings, i think... can you think of an example that wouldn't work, or are we just worried about the variables looking weird? Maybe i'm not grasping it entirely 😕 

OK, so we can do categories and such using the nested approach, but all in a single file. This sounds like something to test out, yeah!

OK so we'd have a URL in the file that points at it's location in the master branch of our Github repo? OK. 

I guess one thing I don't get is how we'd re-integrate the files back into GitHub. Travis integration doesn't do it. Is there a script or bot or app we'd run intermittently to copy translations back into the repo? Maybe: https://docs.transifex.com/projects/downloading-translations#using-the-api-or-command-line-client-to-download-translated-files

Actually a lot of good resources here: https://duckduckgo.com/?q=transifex+PR+github+app&atb=v121-6&ia=software

Although the main integration code is in Java... dunno. Maybe we would need to write a little script to use the CLI client and open a PR once per week or something?

**Comment by gauravano** [comment]
Hey @ebarry, actually that notification is for the test account which I created for PublicLab. I cross-checked the subscription plan of PublicLab account too to be sure and there was no warning there. You received that notification because you are also added as admin for test account. Thanks!

**Comment by gauravano** [comment]
> OK, so we can do categories and such using the nested approach, but all in a single file. This sounds like something to test out, yeah!

Actually, I am not able to understand whether categorization of files currently present at Transifex is automatically done at time of upload or is it done after uploading files to Transifex to organize files in folders replicating the locales folder.
 
I read the @jiteshjha's research notes(final work product and also prepared gists), it seems like all files are uploaded at Transifix first and then categorized(Yes, it can be done, I learnt it from docs). The moto behind using the directories may be to organize the files and for fast searching. Like, if I want to introduce new key-value then instead of carefully traversing the file for perfect place, the key-value can be added fastly using directory structure. 

We can merge all the YAML files into a single using the https://gist.github.com/branch14/1010381, it's old but I tried it and seems to work for us. 

@jywarren please let me know your thoughts about merging into single fine?

Thanks!

**Comment by gauravano** [comment]
Also, there's a blog post regarding the Transifex integration https://blog.ghost.org/transifex-integration/  which is nicely presented.

**Comment by jywarren** [comment]
Ok awesome. Thanks for all this great research. Let's go for a single file
and we can consider changes to this in the more distant future if and when
we find that the file gets unmanageable in size.

Thanks!

**Comment by gauravano** [comment]
Hi @jywarren @ebarry!

#### Status update regarding the Translation project and discussion about next steps:

**1. File Structure change:**
We have successfully converted all the locales in their single language specific file and would upload them to Transifex as soon as https://github.com/publiclab/plots2/pull/4751 get's complete.This PR is raised by a contributor and aims at putting quotes around values to replicate the old config of string-value. It will be merged by next week. 

After that, I would upload the new en.yml file on Transifex and will upload all the language files as their translations. Transifex would be in sync with the file from Git repository for uploading new English strings, and for taking new translations from Transifex, initially, I would raise the PR for some time, I have configured the file .tx file in my system. We can use https://docs.transifex.com/integrations/ruby - this offers a gem (transifex-interface-ruby) for automating the translation flow. We can explore them after this project. What do you think?

**2. Translation Prompts:** 
We can use helper for showing prompts like this - https://github.com/gauravano/plots2/blob/12d5fb8626a9b3e225aad24f02d7017a9a6d5223/app/helpers/application_helper.rb#L152-L162

This way we will be having all the control in our hand for translations. We would be using rails default translate helper in our helper function to get the translated value and will pass the span in the view with prompt and all the styling.

**3. Helper for easing developer to understand:**
![helper_part](https://user-images.githubusercontent.com/20878070/52502978-327f2800-2c0a-11e9-850f-4fd25c4587e3.png)

If we want to pass the value in the translation helper made above then we will have to do our work of fetching the key for each value we receive in translation helper. This work can be done, actually, I have prepared this too - https://github.com/gauravano/plots2/blob/12d5fb8626a9b3e225aad24f02d7017a9a6d5223/app/helpers/application_helper.rb#L164-L170

This way we can get the key for each value we use in the view. And then, after getting the key, we can have the translation from default translation helper and at final step just return the span containing the prompt as output. But, this has a special case, not all the translations are fixed like for some cases we also pass the variables for making the translation complete like count or name passed. So, for these special cases, we would be using the whole value too - 

Example:
```ruby
en:
  dashboard:
    moderate:
      pending_approval: "Pending approval by <a href='%{url}'>community moderators</a>. Please be patient!"
```
The above string has key - dashboard.moderate.pending_approval and value "Pending approval by <a href='%{url}'>community moderators</a>. Please be patient!" 

So, we would be passing the whole value from our controllers/models/views with the above string as a whole, because while finding the key we want it to be exactly same. Are we fine with passing the whole value, containing variables and in some cases this value will be long too.

Please comment your views about 2 and 3 and what approach you think is apt? Thanks!!

**Comment by jywarren** [comment]
Hi, sorry for slow response! 

> automating translation flow 

This sounds great; we might start by version tracking the config file and steps you take just so we have a way to reproduce it. But yeah, very interested to look at some integrations, thanks!

3. this sounds great. I know the key texts may be long, but the idea is to not use it for very long texts (we can do this gradually) and that people should be able to supply an English text for a new string and not have to worry about setting up translations.

One query - what if someone changes the English text, will we have to show them how to change the lookup key as well? Could we auto-detect this with Dangerbot and show people how it works or point them to the corresponding section of the translation files? Maybe we could detect "unreferenced" translations which remain when a key is changed and the translations keys aren't updated to match - and post a Dangerbot notice for any to remind people to update the keys?

**Comment by jywarren** [comment]
For 2., I'd like to enable this for only those who have a URL parameter like `translation=true` but i also [see your comment](https://github.com/publiclab/plots2/pull/4787#issuecomment-462439466) that we'll be changing this gradually, string by string, so I think that's OK too, actually that's probably easier! Thanks @gauravano !!!!

**Comment by gauravano** [comment]
> 1. this sounds great. I know the key texts may be long, but the idea is to not use it for very long texts (we can do this gradually) and that people should be able to supply an English text for a new string and not have to worry about setting up translations.

Can we have a call for Translation helper part, because I think, I would require some more clarity on this? Thanks!

**Comment by gauravano** [comment]
> For 2., I'd like to enable this for only those who have a URL parameter like `translation=true` but i also [see your comment](https://github.com/publiclab/plots2/pull/4787#issuecomment-462439466) that we'll be changing this gradually, string by string, so I think that's OK too, actually that's probably easier!

Ok, working on it.

**Comment by gauravano** [comment]
@jywarren

**Comment by jywarren** [comment]
Hi Gaurav, sorry - I bet it's late there now. I could potentially meet on Monday if you want to name a time?

**Comment by jywarren** [comment]
If there's anything I can help clarify or discuss in the meantime I'm happy to!

**Comment by gauravano** [comment]
Let's focus here a bit: 

![helper_part](https://user-images.githubusercontent.com/20878070/52880919-463a0980-3189-11e9-9d71-1b0514adf833.png)

Currently, we use `<%= t() %>` helper, also known as translate() .

Let's take a example: There's a key  ---> **users.reset.reset_password**  and it has corressponding value ---> "Reset your password".

It is saved in our en.yml file as 
![saved_as](https://user-images.githubusercontent.com/20878070/52881487-b2693d00-318a-11e9-9982-3f40f84cd9de.png)

Now, could you please clarify how the translation helper(translate is alias of t helper only so let's use translation) would work?

**Comment by gauravano** [comment]
@jywarren

**Comment by jywarren** [comment]
Ah, i see. So, the idea is that we'd override the `t()` helper, OR we'd add
a new one, which would look up the string, but if it didn't find it, it'd
use the supplied string as the output. That way, when adding new strings to
the site, new contributors could just add text in English using this
helper, and wouldn't have to know about or deal with adding translation
strings -- we'd pick that up later. The new helper would fallback upon the
supplied string. Does that make sense?

**Comment by gauravano** [comment]
Ok, I think, I get your point. 

Let's say, I am a new developer and want to add a new translation to the website. So, with our current setup, I (a new dev) would add a key and value to the en.yml and will use that translation in code using <%= t() %> and nothing else at all. 

**Translation helper method:**
I will use the method <%= translation("Value ") %> in the code, this will pass the value in the helper instead of the normal flow of passing key. that value will be checked in the en.yml file by comparing it with values and as it's a new translation so will not be found there. So, the value will be returned as it is without any change in case of any language.

Correct @jywarren?

**Comment by gauravano** [comment]
Also, we can have a call today at  11pm (IST) what do you say?

**Comment by jywarren** [comment]
I'm so sorry, as you can probably tell, i'm behind on communication this week... catching up as I can! Yes, you have it just right in your summary. New contributors won't have to touch `en.yml`, and it could be done in follow-up and/or via a script. I'd like to document the sequence here as well, somewhere we can easily point people and simply explain how this works, without overwhelming them. Something like:


```markdown
## Translation

Adding text to the site usually happens in English. But we want our site to be readable by ...

So, all you have to do is wrap your text in this format:

...

This will default to your string of English text. But later, we'll ensure this gets submitted to our community translation system...

```

What do you think?

**Comment by gauravano** [comment]
Ok, I will complete prompt part first and then make a helper for this one. I will add the documentataion too. Also, should I upload the latest en.yml to Transifex or you'll do it? Also, we need to answer this one - https://github.com/publiclab/plots2/issues/4766#issuecomment-460827084 before uploading the file. Thanks!

**Comment by jywarren** [comment]
Can you upload the latest en.yml? Thank you!!!

> @jywarren we use zh-CN locale in our codebase but translations at Transifex are available in zh, so which one we should choose?

I'm not sure here... can you check if zh is traditional or simplified? Whatever is most standard I'm fine with as long as we are aware of the distinction (to my understanding, Taiwan uses traditional and mainland China uses simplified). Thanks!

**Comment by gauravano** [comment]
> Can you upload the latest en.yml?

Will do. We just need to upload the file at Transifex and put it in sync with https://github.com/publiclab/plots2/blob/master/config/locales/en.yml . 

> I'm not sure here... can you check if zh is traditional or simplified? Whatever is most standard I'm fine with as long as we are aware of the distinction (to my understanding, Taiwan uses traditional and mainland China uses simplified). Thanks!

Here's the list of languages by Transifex - https://www.transifex.com/explore/languages/ . Ah, regarding the difference between language codes for zh vs zh-CN, check here - https://en.wikipedia.org/wiki/Chinese_Wikipedia and https://stackoverflow.com/questions/4892372/language-codes-for-simplified-chinese-and-traditional-chinese

Thanks!

**Comment by jywarren** [comment]
How is Transifex's default working for zh-CN vs. zh-TW?

**Comment by gauravano** [comment]
Hey @jywarren,

> How is Transifex's default working for zh-CN vs. zh-TW?
Didn't get your question. I think we can go with zh_CN as what I get from google searches is that zh_TW is used majorly in Taiwan and zh_CN is kind of most used/general. Sorry my knowledge of languages is not so good? I would like to know on which basis you've chosen the languages present on Transifex and also would you be considering zh_CN on basis of population or which factor?


We have 71 files on Transifex. Should we delete them all and upload our new merged locale file?

Thanks!

**Comment by gauravano** [comment]
Check this out too - https://stackoverflow.com/a/9713377/7057132 @jywarren

**Comment by jywarren** [comment]
>  I think we can go with zh_CN as what I get from google searches is that zh_TW is used majorly in Taiwan and zh_CN is kind of most used/general. 

I meant, how does Transifex, in their defaults, address the issue? I think we are OK using primarily `zh_CN`, but we can revisit if there is interest in supporting both, later. Thank you!

Yes, let's delete and merge! Great!

**Comment by gauravano** [comment]
@jywarren transifex added a new method which is awesommmeeee - https://docs.transifex.com/integrations/transifex-github-integration :100:

**Comment by gauravano** [comment]
I am deleting old data and then we can do the setup

**Comment by gauravano** [comment]
@publiclab/community-reps 

I have deleted all the data from transifex. 
@jywarren these are languages present now. Good to continue? Any addition or deletion of language?

English, Arabic, Chinese(zh_CN), German, Hindi, Italian, Korean, Portuguese(Brazil)pt_BR, Spanish(es)

**Comment by gauravano** [comment]
@jywarren I sent the Transifex installation request to plots2. I think we cam install it now. Also, there are 2 join requests pending on Transifex account - 
![transifex_req](https://user-images.githubusercontent.com/20878070/54457695-19791200-4788-11e9-9e55-a5cb2f4a12ed.png)

One is old(2017) and another is by a contributor who told in an issue that they can help in translation so I gave them our account link. So, on which basis we should accept requests?

**Comment by jywarren** [comment]
I guess accept all, no? Approving now!

**Comment by gauravano** [comment]
Have you approved Transifex app install permissions @jywarren ?

**Comment by jywarren** [comment]
yes, i did! did you see?

**Comment by gauravano** [comment]
I tried many times but it's not showing transifex app installed :point_down: 
![Screenshot from 2019-03-19 21-30-32](https://user-images.githubusercontent.com/20878070/54621777-51d16680-4a8e-11e9-82d4-0ab7e04e11e8.png)

Could you go at https://www.transifex.com/publiclab/settings/integrations/ and install the transifex integration? 

Thanks!
@jywarren

**Comment by jywarren** [comment]
Trying, not sure what's going on but it looked installed when i did it - i
"accepted" it again today!

**Comment by jywarren** [comment]
Ah, it had a second step, allowing TF to post on my behalf. We may need to
do this with the Plotsbot or another dummy GitHub account, instead of it
being posted as me? Worth testing out!

**Comment by gauravano** [comment]
ok, I see authorized now 
![Screenshot from 2019-03-20 00-57-46](https://user-images.githubusercontent.com/20878070/54635948-56584800-4aab-11e9-87e3-5ca0cb6e2420.png)

**Comment by gauravano** [comment]
@jywarren refer this https://docs.transifex.com/integrations/transifex-github-integration

**Comment by gauravano** [comment]
they have bot - 
![Screenshot from 2019-03-20 01-00-09](https://user-images.githubusercontent.com/20878070/54636054-8b649a80-4aab-11e9-8a7f-3fc9095d0c10.png)

**Comment by jywarren** [comment]
Great!

**Comment by gauravano** [comment]
I did the remaining setup but the sad part of this integration is we can't make good use of this until we reach 100% reviewed status on a language. I have dropped a suggestion to Transifex regrading this - https://community.transifex.com/t/transifex-should-be-able-to-push-translations-more-often-transifex-github-integration/261 

so anyone of us can raise PR weekly till we reach the 100% reviewed status for all langs and after that this integration will raise PR for each reviewed string addition(I will document the steps for PR)?

**Comment by jywarren** [comment]
Hmm, weird, too bad! But yes, until then that sounds good!

**Comment by jywarren** [comment]
Hi @gauravano shall we update this and wrap it up? I think we are basically almost there, no? Thank you!!!

**Comment by gauravano** [comment]
Hi @jywarren, yes, we are almost done. Give me a day to create a checklist and I guess, I will update the research notes posted by Liz, we can then, start using multiple languages.

Thanks!

**Comment by jywarren** [comment]
awesome!!

**Comment by jywarren** [comment]
Sorry about the stalebot message here, it was a mistake! 😅
